File: app/src/main/java/ua/in/quireg/anothermovieapp/activities/MainActivity.java
package ua.in.quireg.anothermovieapp.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import ua.in.quireg.anothermovieapp.R;
import ua.in.quireg.anothermovieapp.common.Constants;
import ua.in.quireg.anothermovieapp.interfaces.OnFragmentInteractionListener;
import ua.in.quireg.anothermovieapp.core.MovieItem;
import ua.in.quireg.anothermovieapp.services.SyncMovieService;
import ua.in.quireg.anothermovieapp.ui.FavouritesMoviesGridViewFragment;
import ua.in.quireg.anothermovieapp.ui.MoviesGridViewFragment;
import ua.in.quireg.anothermovieapp.ui.PopularMoviesGridViewFragment;
import ua.in.quireg.anothermovieapp.ui.TopRatedMoviesGridViewFragment;

public class MainActivity extends AppCompatActivity implements OnFragmentInteractionListener {

    private static final String LOG_TAG = MainActivity.class.getSimpleName();

    private static final int POPULAR_MOVIES_TAB_POSITION = 0;
    private static final int TOP_RATED_MOVIES_TAB_POSITION = 1;
    private static final int FAVOURITE_MOVIES_TAB_POSITION = 2;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_ama_main_activity);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        //Set title to "Popular" as it is the first fragment to be shown.
        getSupportActionBar().setTitle(getResources().getString(R.string.popular_tab_name));
        SectionsPagerAdapter mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager(), this.getApplicationContext());

        // Set up the ViewPager with the sections adapter.
        ViewPager mViewPager = (ViewPager) findViewById(R.id.container);
        mViewPager.setAdapter(mSectionsPagerAdapter);

        TabLayout tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(mViewPager);

    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_ama_main_activity, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        } else if (id == R.id.action_refresh) {
            SyncMovieService.startActionFetchMovies(getApplicationContext(), Constants.POPULAR, "1");
            SyncMovieService.startActionFetchMovies(getApplicationContext(), Constants.TOP_RATED, "1");
            return true;
        }

        return super.onOptionsItemSelected(item);
    }


    @Override
    public void onFragmentMessage(String TAG, Object data) {
        Intent i = new Intent(MainActivity.this, MovieActivity.class);
        Bundle args = new Bundle();
        args.putSerializable(Constants.MOVIE, (MovieItem)data);
        i.putExtras(args);
        startActivity(i);
    }


    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentPagerAdapter {

        private Context mContext;
        private MoviesGridViewFragment popularMoviesFragment;
        private MoviesGridViewFragment topRatedMoviesFragment;
        private MoviesGridViewFragment favouritesMoviesFragment;


        public SectionsPagerAdapter(FragmentManager fm, Context context) {
            super(fm);
            this.mContext = context;
        }

        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case POPULAR_MOVIES_TAB_POSITION:
                    if(popularMoviesFragment == null) {
                        popularMoviesFragment = new PopularMoviesGridViewFragment();
                        Bundle args = new Bundle();
                        args.putSerializable(Constants.FRAGMENT_TAG,Constants.POPULAR);
                        popularMoviesFragment.setArguments(args);

                    }
                    return popularMoviesFragment;
                case TOP_RATED_MOVIES_TAB_POSITION:
                    if(topRatedMoviesFragment == null) {
                        topRatedMoviesFragment = new TopRatedMoviesGridViewFragment();
                        Bundle args = new Bundle();
                        args.putSerializable(Constants.FRAGMENT_TAG,Constants.TOP_RATED);
                        topRatedMoviesFragment.setArguments(args);
                    }
                    return topRatedMoviesFragment;
                case FAVOURITE_MOVIES_TAB_POSITION:
                    if(favouritesMoviesFragment == null) {
                        favouritesMoviesFragment = new FavouritesMoviesGridViewFragment();
                        Bundle args = new Bundle();
                        args.putSerializable(Constants.FRAGMENT_TAG, Constants.FAVOURITES);
                        favouritesMoviesFragment.setArguments(args);
                    }
                    return favouritesMoviesFragment;
            }

            return null;
        }

        @Override
        public int getCount() {
            return 3;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return mContext.getResources().getString(R.string.popular_tab_name);
                case 1:
                    return mContext.getResources().getString(R.string.top_rated_tab_name);
                case 2:
                    return mContext.getResources().getString(R.string.favourites_tab_name);
            }
            return null;
        }
    }
}

File: app/src/main/java/com/anothermovieapp/view/MainActivity.kt
/*
 * Created by Arcturus Mengsk
 *   2021.
 */

package com.anothermovieapp.view

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.anothermovieapp.R
import com.anothermovieapp.common.Constants
import com.anothermovieapp.repository.EntityDBMovie
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.tabs.TabLayout
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {
    lateinit var mSectionsPagerAdapter: SectionsPagerAdapter
    var popularMoviesFragment: MoviesGridViewFragment? = null
    var topRatedMoviesFragment: MoviesGridViewFragment? = null
    var favouritesMoviesFragment: MoviesGridViewFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val rootView = LayoutInflater.from(this)
            .inflate(R.layout.activity_ama_main_activity, null)

        setContentView(rootView)
        rootView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.setTitleTextColor(resources.getColor(R.color.primaryColor, theme))
        setSupportActionBar(toolbar)
        val appBarLayout = findViewById<AppBarLayout>(R.id.appbar)
        // Set up the ViewPager with the sections adapter.
        val viewPager = findViewById<ViewPager>(R.id.container)
        popularMoviesFragment =
            supportFragmentManager.findFragmentByTag(
                "android:switcher:" + viewPager.id + ":" + POPULAR_MOVIES_TAB_POSITION
            )
                    as MoviesGridViewFragment?
        if (popularMoviesFragment == null) {
            popularMoviesFragment = PopularMoviesGridViewFragment()
        }
        popularMoviesFragment!!.fragmentListener(fragmentListener)
        topRatedMoviesFragment =
            supportFragmentManager.findFragmentByTag(
                "android:switcher:" + viewPager.id + ":" + TOP_RATED_MOVIES_TAB_POSITION
            )
                    as MoviesGridViewFragment?
        if (topRatedMoviesFragment == null) {
            topRatedMoviesFragment = TopRatedMoviesGridViewFragment()
        }
        topRatedMoviesFragment!!.fragmentListener(fragmentListener)
        favouritesMoviesFragment =
            supportFragmentManager.findFragmentByTag(
                "android:switcher:" + viewPager.id + ":" + FAVOURITE_MOVIES_TAB_POSITION
            )
                    as MoviesGridViewFragment?
        if (favouritesMoviesFragment == null) {
            favouritesMoviesFragment = FavouritesMoviesGridViewFragment()
        }
        favouritesMoviesFragment!!.fragmentListener(fragmentListener)
        mSectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager)


        viewPager.adapter = mSectionsPagerAdapter
        val tabLayout = findViewById<TabLayout>(R.id.tabs)
        tabLayout.setupWithViewPager(viewPager)
//        ViewCompat.setOnApplyWindowInsetsListener(rootView) { v: View, insets: WindowInsetsCompat ->
//            ViewCompat.setOnApplyWindowInsetsListener(rootView, null)
//            appBarLayout.setPadding(
//                    v.paddingLeft,
//                    insets.systemWindowInsetTop,
//                    v.paddingRight,
//                    v.paddingBottom)
//            mSystemInsetTop = insets.systemWindowInsetTop
//            mSystemInsetBottom = insets.systemWindowInsetBottom
//            val pageNumber = findViewById<TextView>(R.id.pageNumberAndTotal)
//            val lp = pageNumber.layoutParams as FrameLayout.LayoutParams
//            lp.bottomMargin += insets.systemWindowInsetBottom
//            pageNumber.layoutParams = lp
//            val pb = findViewById<ProgressBar>(R.id.progressBar)
//            val lp2 = pageNumber.layoutParams as FrameLayout.LayoutParams
//            lp.bottomMargin += insets.systemWindowInsetBottom
//            pb.layoutParams = lp2
//            insets.consumeSystemWindowInsets()
//        }

        val pageNumberAndTotal : TextView = rootView.findViewById(R.id.pageNumberAndTotal)
        val lp= pageNumberAndTotal.layoutParams as FrameLayout.LayoutParams
        lp.bottomMargin = lp.bottomMargin + getNavBarHeight()
        pageNumberAndTotal.layoutParams = lp
        rootView.invalidate()
    }

    open fun getNavBarHeight(): Int {
        var result = 0
        val resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android")
        if (resourceId > 0) {
            result = resources.getDimensionPixelSize(resourceId)
        }
        return result
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_ama_main_activity, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            //            R.id.action_settings -> {
            //                return true
            //            }
            R.id.action_refresh -> {
                popularMoviesFragment?.reload()
                topRatedMoviesFragment?.reload()
                favouritesMoviesFragment?.reload()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    val fragmentListener = object : OnFragmentInteractionListener {
        override fun onFragmentMessage(TAG: String?, data: Any?) {
            val i = Intent(this@MainActivity, DetailsActivity::class.java)
            val args = Bundle()
            args.putSerializable(Constants.MOVIE, data as EntityDBMovie?)
            i.putExtras(args)
            startActivity(i)
        }
    }

    inner class SectionsPagerAdapter internal constructor(
        fm: FragmentManager
    ) : FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        override fun getItem(position: Int): Fragment {
            return when (position) {
                POPULAR_MOVIES_TAB_POSITION -> {
                    popularMoviesFragment!!
                }
                TOP_RATED_MOVIES_TAB_POSITION -> {
                    topRatedMoviesFragment!!
                }
                else -> {
                    favouritesMoviesFragment!!
                }
            }
        }

        override fun getCount(): Int {
            return 3
        }
    }

    companion object {
        private const val TOP_RATED_MOVIES_TAB_POSITION = 1
        private const val FAVOURITE_MOVIES_TAB_POSITION = 2
        private const val POPULAR_MOVIES_TAB_POSITION = 0
    }
}